pallets/collective-proxy/src/lib.rs

```diff
@@ -121,5 +178,66 @@ pub mod pallet {
 
             Ok(())
         }
+
+        /// Register a proxy account for the sender that is able to make calls on its behalf.
+        ///
+        /// The dispatch origin for this call must be _Signed_.
+        ///
+        /// Parameters:
+        /// - `proxy`: The account that the `caller` would like to make a proxy.
+        /// - `filter`: Call filter used for the proxy
+        #[pallet::call_index(1)]
+        #[pallet::weight(T::WeightInfo::add_proxy(T::MaxProxies::get()))]
+        pub fn add_proxy(
+            origin: OriginFor<T>,
+            proxy: T::AccountId,
+            filter: T::CallFilter,
+        ) -> DispatchResult {
+            T::CollectiveProxy::ensure_origin(origin)?;
```

Comment:
The collective proxy origin can register arbitrary account.

Security-wise, what do you think about this?

Reply:
Yes, true. It's a security flaw. AFAIU the current pallet design also has it, right? In order to address it, the major change of the pallet needed. Proxy pallet uses announcements mechanism to provide the ability to cancel "proxying". So it's also the tradeoff: should we keep current minimalist design for the pallet or we make it more secure. I assumed that my task was in the scope of the first approach.

Reply:
Maybe we're not thinking about the same flaw :)

The current pallet's custom `CollectiveProxy` origin filter is hardcoded in the runtime. Runtime upgrade is a heavily privileged operation, and cannot be done easily.

On the other hand, in this new extrinsic call, the collective proxy origin can register any new filter they want.
E.g. they can decide to create a proxy to my own account, and execute any action they want (based on the filter). No one would like that 🙂.

Reply:
True, true. So bringing announcements into the design? Are there any other means to prevent it?

Reply:
Having an announcement delay is good feature.
Keep in mind though that this pallet is used currently only with _private_ accounts, so even if we had announcements, it wouldn't improve the current situation.

For the suggestions on how to improve this, I have two:
1. Keeping your approach, introduce a special origin type that can register a new proxy. E.g. make it root-only.
2. Taking a different approach, expand the initial pallet code to allow registering (or _defining_) multiple account Ids & filters. E.g. something like a `type` that implements `Get<Vec<(Origin, AccountId, CallFilter)>`. That way the same security level as it is now is kept.

Reply:
Implemented the first approach. Corresponding test also added.